### ISSUE TITLE
Reliable timer ticks

### ISSUE THREAD

**Issue by amelchio** [home-assistant/architecture#50]
The current timer tick code has some issues.

__1) Double events__

The timer tries to tick a whole number of seconds since it was started. Thus, if it is started very close to a second rolling over, many ticks will fire on the wrong second. This is because we are usually a little bit late with scheduled tasks. I added a log statement to show this:

```
2018-07-29 09:27:16 [...] Timer tick 2018-07-29 07:27:16.998479+00:00
2018-07-29 09:27:18 [...] Timer tick 2018-07-29 07:27:18.000322+00:00 <-- :17
2018-07-29 09:27:18 [...] Timer tick 2018-07-29 07:27:18.995209+00:00
2018-07-29 09:27:20 [...] Timer tick 2018-07-29 07:27:20.000136+00:00 <-- :19
2018-07-29 09:27:21 [...] Timer tick 2018-07-29 07:27:21.000139+00:00 <-- :20
2018-07-29 09:27:22 [...] Timer tick 2018-07-29 07:27:22.000105+00:00 <-- :21
2018-07-29 09:27:22 [...] Timer tick 2018-07-29 07:27:22.999665+00:00
```

__2) Missing events__

The code detects some timer sync issues and resolves them by skipping events. This can cause time-based automations to not fire.

__3) Confusing error messages__

We log an error when the timer gets out of sync. This helps to find bad code (which spends too long in the event loop) but there are a lot of cases where nothing is actually wrong. Rather, slow hardware (like a Raspberry Pi) and/or the [poor scheduling in Python](https://forum.dabeaz.com/t/thats-a-nice-event-loop-itd-be-a-shame/118) causes the timer code to just not run in time because we were busy doing legitimate work.

The forum has many threads asking about this issue and including those that did not create a forum thread, an enormous amount of time must be wasted wondering about that error message. 

Because of 2) I am hesitant to tell users that they should only worry about the message if they notice other problems.

## Proposal

During my investigations I have created [a WIP branch](https://github.com/amelchio/home-assistant/compare/dev...synthesize-timer-tick) that solves most of this by trying to tick when a second rolls over and by injecting the missed time-changed events when they are detected. Out-of-sync errors are only reported when they are truly severe.

I don't think my branch is perfect yet so I have created this issue to discuss how to best handle the above points.

**Comment by pvizeli** [comment]
The time handling was create before we use asyncio. With asyncio we have a time sheduler that work with best effort.

I think we don't need now a own scheduler. We should migrate the code to use the internal asyncio scheduler. That should also speedup the core while we don't need run this timechange event every seconds.

**Comment by balloob** [comment]
When a timer goes out of sync, there is bad code or someone's running too much and their system is resource constrained. We only fire out of sync if we are called a second after our scheduled time, that's a lot of clock cycles!

I agree with Pascal that using asyncio scheduling should be used. However, we'll need to find a good way to flush pending callbacks in tests.

Missing events suck, we might as a temporary measure schedule the tick to start at 0.0005 of a second.

I'm hesitant to just start firing all missed events, it will cause more work instead of allowing the system to catch up on breath, taking it into a downwards spiral. We should come up with a recovery mode

**Comment by balloob** [comment]
Btw I do like the time changed events as it ensures the full system is event driven (it's called event sourcing). However, we don't store them either so already skipped on that

**Comment by balloob** [comment]
Yeah, dropping it would be fine. For testing however, it would be nice if there was something like Trio's [MockClock](http://trio.readthedocs.io/en/latest/reference-testing.html#time-and-timeouts)

**Comment by amelchio** [comment]
I now implemented that and it solves all three issues 👍. However, I am indeed stuck on how to get the tests working again.

Also, the flexible time patterns made me unable to calculate the duration until the next matching time. For example, `minutes: /59` will fire 120 seconds in a row and then pause for 3480 seconds. I tried some optimizations but it gets messy real quick.

Link: [My WIP branch](https://github.com/home-assistant/home-assistant/compare/dev...amelchio:tickless) (mainly the first two commits)

**Comment by pvizeli** [comment]
Here are good example to calc format like this and more very simple:
https://github.com/gawel/aiocron

**Comment by balloob** [comment]
If a computer goes in standby and comes back up a day later, we should not fire 86400 events, so we still need some reset threshold. Really if a system has missed 2 ticks, it should already be looked at.

**Comment by amelchio** [comment]
I have added code to handle the standby case but please note that my branch is just a Proof of Concept. I do not have time to work on the tests.

As for two missed ticks being something that should be looked at, I actually agree. However, please see my point 3) above. The problem is, fittingly, a lack of time. The few people that fully understand the current error message are not able to investigate all cases.

This is on top of the fact that the poor scheduling in Python can make the error appear even when the only fix is "don't do that". Things like purging a huge SQLite database, adding a sensor that does several seconds of calculations or just having a lot of components doing setup can bring things down.

With the above in mind, I think it is a disservice to print a prominent error message in this situation. It is often just a waste of people's time.

My fix is to make the situation mostly harmless (by ensuring that handlers always execute) and then downgrade or remove the error message. Users can then ignore the issue unless they are indeed affected.

I do think the error message has value, it has helped me fix several issues. I just think we should move it to a dedicated component that people can enable if they actually notice a performance problem. This would allow us to expand the diagnostics and maybe add statistics that could give some indication about the extent of the problem.

**Comment by OttoWinter** [comment]
Great work! I would like to point out one limitation with purely using the eventloop: Up until Python 3.7.1 [`call_later`](https://docs.python.org/3/library/asyncio-eventloop.html#asyncio.loop.call_later) only accepts delays of up to one day. I think for a home automation system with usually a bit higher delays this could become a problem. For example if the time trigger would accept day-of-week arguments (https://github.com/home-assistant/architecture/issues/88#issuecomment-427034406)

I personally would say it would be easier to just build our own scheduler system. This would have a few key features:

 * Ability to have delays longer than a day
 * More access to scheduled tasks from the test environment.
 * We could even have the new scheduler system use the existing `time_changed` event. We would have one global listener attached to `time_changed` checking for which tasks to run next from a heap. This would make test migration a lot easier.
 * Potentially: task priority

This way, we could also have a slow rolling migration to remove `time_changed`. If the scheduler knows that the next task to run only is in 5 minutes, it could wait for 5 minutes until popping the next element off the heap.

In the end, the base event loop asyncio implementation of  [`call_at`](https://github.com/python/cpython/blob/d5bd036138881bb90a803397d992870a46fbdc2d/Lib/asyncio/base_events.py#L660) only adds the task to an internal priority queue anyway, so there would not really be a big difference. Plus I think the implementation wouldn't be too complicated, 50 lines or so max

**Comment by pvizeli** [comment]
I would prefer `call_at` because they already exists and can improve with system like uvloop and is well tested. We use actual also the loop for schedule, so if we want use a own, we don't need change it.

The new smart calculation should allow us to simple use `call_at`.

**Comment by balloob** [comment]
The problem with using just `call_at` is that we can't have our tests flush it and run all pending handlers.

**Comment by amelchio** [comment]
I intend to close this issue because I think @OttoWinter has handled it very well with home-assistant/home-assistant#17199 ... even though the timer ticks are still not reliable, we no longer expect them to be (because we no longer use equality for comparisons).

Before I do, though, I am going to reiterate that I would love for the "Timer got out of sync" `error` to be downgraded, preferably all the way down to `debug`. I my view (and after Otto's PR) that's all it is: a debugging tool.

**Comment by balloob** [comment]
I agree. I am fine with a PR for that, but then let's replace it with a TIMER_OUT_OF_SYNC event. That way if we ever build a find problems component, we can look for those events.
